Slim/Networking/SimpleAsyncWS.pm

```diff
@@ -0,0 +1,385 @@
+	my ($self) = @_;
+
+	$self->_receive(1);
+
+	return;
+}
+
+
+sub receiveSync {
+	my ($self, $timeout, $cbRead, $cbReadFailed ) = @_;
+
+	main::INFOLOG && $log->is_info && $log->info("Single Receive sync timeout : $timeout");
+
+	#Set callack on object
+	$self->{cb_Read} = $cbRead;
+	$self->{cb_Read_Failed} = $cbReadFailed;
+	$self->{continue_listening} = 0;
+
+	$self->_receive(0, $timeout);
+
+	return;
+}
+
+
+sub _receive {
+	my ($self, $isAsync, $timeout) = @_;
+	main::DEBUGLOG && $log->is_debug && $log->debug("Starting Listening");
+
+	#Operation to check on the socket reading
+	#if isAsync is true it will check in a non-blocking way and initiate a future check asynchronously for continuose listening.
+	#if isAync is false it will wait for something to arrive for $timeout length. This is blocking so suggest this is < 1 second.  This is a single read.
+
+	if ( !$isAsync || ($isAsync && $self->{continue_listening}) ) {
+
+		my $s = IO::Select->new();
+		$s->add($self->{tcp_socket});
+		$! = 0;
+
+		main::DEBUGLOG && $log->is_debug && $log->debug("Checking the socket for something to read");
+		my @ready = $isAsync ? $s->can_read(0) : $s->can_read($timeout);
+
+		if (@ready) {
+			my $recv_data;
+			my $bytes_read = sysread $ready[0], $recv_data, 16384;
+			if (!defined $bytes_read) {
+
+				$log->error("Error reading from socket : $!");
+				$self->{cb_Read_Failed}->();
+
+				# poll again in 1 second
+				$self->_continueListen(1) if $isAsync;
+
+			} elsif ($bytes_read == 0) {
+
+				# Remote socket closed
+				$log->Error("Connection terminated by remote. $!");
+
+				$self->{cb_Read_Failed}->();
+
+				# We will not continue (if ASync)
+				$self->{continue_listening} = 0;
+
+			} else {
+
+				main::DEBUGLOG && $log->is_debug && $log->debug("Received data : $recv_data ");
+				$self->{client}->read($recv_data);
+
+				# if Async, poll immediately so that we pull everything off the socket if something is there.
+				$self->_continueListen(0) if $isAsync;
+
+			}
+
+		} else {
+
+			main::DEBUGLOG && $log->is_debug && $log->debug("No Data Present, continue listening");
+
+			# poll again in 1 second
+			$self->_continueListen(1) if $isAsync;
+
+		}
+	}
+
+	return;
+}
+
+
+sub _continueListen {
+	my ($self, $pollTimeSeconds) = @_;
+
+	Slim::Utils::Timers::setTimer($self, time() + $pollTimeSeconds, \&_receiveAsync);
+
+	return;
+}
+
+
+sub send {
+	my ($self, $buf) = @_;
+
+	main::INFOLOG && $log->is_info && $log->info("Sending on web socket : $buf ");
+	$self->{client}->write($buf);
+
+	return;
+}
+
+1;
+
+__END__
+
+=head1 NAME
+
+Slim::Networking::SimpleAsyncWS - Simple WS Client with asynchronous non-blocking socket listening
+
+=head1 SYNOPSIS
+
+use Slim::Networking::SimpleAsyncWS
+
+sub exampleErrorCallback {  
+
+    print("Oh no! An error!\n");
+}
+
+sub exampleWeAreConnected {  
+
+    print("We are connected");
+}
+
+sub exampleCallback {
+    my $buf = shift;   
+
+    print("Got the message.\n");
+	print($buf);
+}
+
```

Comment:
Oh, one nitpick: could you please make sure you're using tabs (not spaces) consistently? And trim trailing spaces. Thanks!

Reply:
Oh yes, I left some trailing spaces in the example.  I'll fix that.

Reply:
Although to be fair to me, they were already in the example that I copied over from SimpleAsyncHTTP that I based it on ;-)

Reply:
Haha... I tend not to change whitespace for the sake of changing it, but would do so when editing the code anyway. With new code we can start clean, even if it's a copy of existing code 😉.